src/libtego_ui/libtego_callbacks.cpp

```diff
@@ -0,0 +1,459 @@
+
+        char serviceIdRaw[TEGO_V3_ONION_SERVICE_ID_SIZE] = {0};
+        tego_v3_onion_service_id_to_string(serviceId.get(), serviceIdRaw, sizeof(serviceIdRaw), tego::throw_on_error());
+
+        auto contactId = QString::fromUtf8(serviceIdRaw, TEGO_V3_ONION_SERVICE_ID_LENGTH);
+        return contactId;
+    }
+
+    // converts the our tego_user_id_t to ricochet's contactId in the form ricochet:serviceidserviceidserviceid...
+    QString tegoUserIdToContactId(const tego_user_id_t* user)
+    {
+        return serviceIdToContactId(tegoUserIdToServiceId(user));
+    }
+
+    shims::ContactUser* contactUserFromContactId(const QString& contactId)
+    {
+        auto userIdentity = shims::UserIdentity::userIdentity;
+        auto contactsManager = userIdentity->getContacts();
+
+        auto contactUser = contactsManager->getShimContactByContactId(contactId);
+        return contactUser;
+    }
+
+    //
+    // libtego callbacks
+    //
+
+    void on_tor_error_occurred(
+        tego_context_t*,
+        tego_tor_error_origin_t origin,
+        const tego_error_t* error)
+    {
+        // route the error message to the appropriate component
+        QString errorMsg = tego_error_get_message(error);
+        logger::println("tor error : {}", errorMsg);
+        push_task([=]() -> void
+        {
+            switch(origin)
+            {
+                case tego_tor_error_origin_control:
+                {
+                    shims::TorControl::torControl->setErrorMessage(errorMsg);
+                }
+                break;
+                case tego_tor_error_origin_manager:
+                {
+                    shims::TorManager::torManager->setErrorMessage(errorMsg);
+                }
+                break;
+            }
+        });
+    }
+
+    void on_update_tor_daemon_config_succeeded(
+        tego_context_t*,
+        tego_bool_t success)
+    {
+        push_task([=]() -> void
+        {
+            logger::println("tor daemon config succeeded : {}", success);
+            auto torControl = shims::TorControl::torControl;
+            if (torControl->m_setConfigurationCommand != nullptr)
+            {
+                torControl->m_setConfigurationCommand->onFinished(success);
+                torControl->m_setConfigurationCommand = nullptr;
+            }
+        });
+    }
+
+    void on_tor_control_status_changed(
+        tego_context_t*,
+        tego_tor_control_status_t status)
+    {
+        push_task([=]() -> void
+        {
+            logger::println("new control status : {}", status);
+            shims::TorControl::torControl->setStatus(static_cast<shims::TorControl::Status>(status));
+        });
+    }
+
+    void on_tor_process_status_changed(
+        tego_context_t*,
+        tego_tor_process_status_t status)
+    {
+        push_task([=]() -> void
+        {
+            logger::println("new process status : {}", status);
+            auto torManger = shims::TorManager::torManager;
```

Comment:
i'm going to assume this is meant to be `torManager`

Reply:
yep